src/sqlfluff/rules/L054.py

```diff
@@ -0,0 +1,111 @@
+"""Implementation of Rule L054."""
+from typing import Optional
+
+from sqlfluff.core.rules.base import BaseRule, LintResult, RuleContext
+
+
+class Rule_L054(BaseRule):
+    """Inconsistent column references in GROUP BY/ORDER BY clauses.
+
+    | **Anti-pattern**
+    | A mix of implicit and explicit column references are used in a GROUP BY clause.
+
+    .. code-block:: sql
+       :force:
+
+        SELECT
+            foo,
+            bar,
+            sum(baz) AS sum_value
+        FROM fake_table
+        GROUP BY
+            foo, 2;
+
+        -- The same also applies to column
+        -- references in ORDER BY clauses.
+
+        SELECT
+            foo,
+            bar
+        FROM fake_table
+        ORDER BY
+            1, bar;
+
+    | **Best practice**
+    | Reference all GROUP BY/ORDER BY columns either by name or by position.
+
+    .. code-block:: sql
+       :force:
+
+        -- GROUP BY: Explicit
+        SELECT
+            foo,
+            bar,
+            sum(baz) AS sum_value
+        FROM fake_table
+        GROUP BY
+            foo, bar;
+
+        -- ORDER BY: Explicit
+        SELECT
+            foo,
+            bar
+        FROM fake_table
+        ORDER BY
+            foo, bar;
+
+        -- GROUP BY: Implicit
+        SELECT
+            foo,
+            bar,
+            sum(baz) AS sum_value
+        FROM fake_table
+        GROUP BY
+            1, 2;
+
+        -- ORDER BY: Implicit
+        SELECT
+            foo,
+            bar
+        FROM fake_table
+        ORDER BY
+            1, 2;
+    """
+
+    config_keywords = ["group_by_and_order_by_style"]
+
+    def _eval(self, context: RuleContext) -> Optional[LintResult]:
+        """Inconsistent column references in GROUP BY/ORDER BY clauses."""
+        # Config type hints
+        self.group_by_and_order_by_style: str
+
+        # We only care about GROUP BY/ORDER BY clauses.
+        if context.segment.name not in {"GroupByClauseSegment", "OrderByClauseSegment"}:
```

Comment:
Rules usually use `is_type()` instead, something like:
```
context.segment.is_type("groupby_clause", "orderby_clause")
```

I think they're functionally identical, but I'd prefer we change this to be consistent with other rules.

Reply:
will do

Reply:
used the types now as you said, I still use the names in constructing `column_reference_category_set` as the numeric_literal segment type is `raw` which is unhelpful. I've made a note in the code